fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/CorrelatedPredicateRewriter.java

```diff
@@ -0,0 +1,126 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+package com.starrocks.sql.optimizer.rewrite;
+
+import com.google.common.collect.Maps;
+import com.starrocks.sql.optimizer.OptimizerContext;
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CaseWhenOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CastOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * used to extract distinct expr which only contains inner table column and replace it with a new columnRef.
+ * e.g. t1 is inner table, t2 is outer table. correlated predicate is:
+ * t1.col1 + t2.col1 = abs(t1.col2) + t2.col1 + concat(t1.col1, t2.col1)
+ * The rewriter will extract abs(t1.col2), t2.col1 and build scalarOperator to new columnRef map like:
+ * t1.col1 : t1.col1
+ * abs(t1.col2) : columnRef1
+ *
+ * then and rewrite the predicate like:
+ * t1.col1 + t2.col1 = columnRef1 + t2.col1 + concat(t1.col1, t2.col1)
```

Comment:
why only handle abs(t1.col2), not handle concat(t1.col1, t2.col1)?

Reply:
Our main purpose is extract all distinct exprs belong to inner table. concat(t1.col1, t2.col1) contains outer table t1.col1. In this expr, the atom expr from inner table is t2.col1, so we only need extract t2.col1.